src/reactA11yAccessibleHeadingsRule.ts

```diff
@@ -0,0 +1,194 @@
+const EMPTY_HEADING_FAILURE_STRING: string = 'Heading elements must not be empty';
+const BAD_HEADING_LENGTH_STRING: string = 'Heading content should be concise';
+const BAD_NUMBER_H1_HEADING_FAILURE_STRING: string = 'H1 heading cannot exceed 2 elements';
+const VALID_HEADING_TYPES: Set<string> = new Set(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);
+const MAX_NUMBER_OF_H1_HEADINGS: number = 2;
+const MAX_HEADING_LENGTH_ATTRIBUTE_NAME: string = 'maxHeadingLength';
+
+interface IValidationOptions {
+    validateText?: boolean;
+    validateOrder?: boolean;
+    validateH1Count?: boolean;
+}
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: ExtendedMetadata = {
+        ruleName: 'react-a11y-accessible-headings',
+        type: 'functionality',
+        description:
+            'For accessibility of your website, there should be no more than 2 H1 heading elements, HTML heading elements must be concise, used for structuring information on the page and non-empty.',
+        options: {
+            maxHeadingLength: 'number'
+        },
+        optionsDescription: 'Heading elements maximum text length',
+        typescriptOnly: true,
+        issueClass: 'Non-SDL',
+        issueType: 'Warning',
+        severity: 'Moderate',
+        level: 'Opportunity for Excellence',
+        group: 'Accessibility'
+    };
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        if (sourceFile.languageVariant === ts.LanguageVariant.JSX) {
+            return this.applyWithWalker(new ReactA11yAccessibleHeadingsWalker(sourceFile, this.getOptions()));
+        } else {
+            return [];
+        }
+    }
+}
+
+class ReactA11yAccessibleHeadingsWalker extends Lint.RuleWalker {
+    //private headings: { [id: string] : IHeadingResult; } = {};
+    private lastKnownHeadingNumber: number | undefined;
+    private numberOfH1Headings: number = 0;
+    private readonly maxHeadingTextLength: number | undefined;
+
+    constructor(sourceFile: ts.SourceFile, options: Lint.IOptions) {
+        super(sourceFile, options);
+        const option = options.ruleArguments.find(a => a.hasOwnProperty(MAX_HEADING_LENGTH_ATTRIBUTE_NAME));
+        if (option && typeof option[MAX_HEADING_LENGTH_ATTRIBUTE_NAME] === 'number') {
+            this.maxHeadingTextLength = option[MAX_HEADING_LENGTH_ATTRIBUTE_NAME];
+        }
+    }
+
+    protected visitJsxSelfClosingElement(node: ts.JsxSelfClosingElement): void {
+        this.validateSelfClosingElement(node);
+        super.visitJsxSelfClosingElement(node);
+    }
+
+    protected visitVariableDeclaration(node: ts.VariableDeclaration): void {
+        const options: IValidationOptions = {
+            validateText: true,
+            validateOrder: false,
+            validateH1Count: false
+        };
+
+        let element = this.tryGetNode(node, ts.SyntaxKind.JsxElement);
```

Comment:
Huh, interesting. Does the rule not normally visit `JsxElement`s? I'm curious if & why you needed to add this?

Reply:
Using `VisitVariableDeclaration` and then traversing `Jsx` child nodes using `this.tryGetNode`. How would you capture the context otherwise?